[Access] Observable interface for unstaked follower engine

smnzhu opened the pull request:
Implements a new Observable for the stream of finalized blocks.

closes https://github.com/dapperlabs/flow-go/issues/5688

smnzhu commented:
@vishalchangrani So since it turns out We don't need the block proposals and we Only need to be notified of finalized blocks, this simplifies things a bit and we actually *can* put many of the modules into a single `Build` method.

vishalchangrani commented on cmd/access/main.go:
```diff
@@ -10,72 +10,54 @@ import (
-		flags.BoolVar(&anb.staked, "staked", defaultConfig.staked, "whether this node is a staked access node or not")
-		flags.StringVar(&anb.stakedAccessNodeIDHex, "staked-access-node-id", defaultConfig.stakedAccessNodeIDHex, "the node ID of the upstream staked access node if this is an unstaked access node")
-		flags.StringVar(&anb.unstakedNetworkBindAddr, "unstaked-bind-addr", defaultConfig.unstakedNetworkBindAddr, "address to bind on for the unstaked network")
+		flags.UintVar(&anb.receiptLimit, "receipt-limit", anb.AccessNodeConfig.receiptLimit, "maximum number of execution receipts in the memory pool")
```
I defined the deaultConfig to make this code read a little better (defaultConfig for all default values).
Also, here we can replace all `anb.XXX.YYY` with `anb.YYY`.

smnzhu replied:
An I see, the reason I changed it is because I moved the defaultConfig to the constructor method. By the time `anb.ExtraFlags` is called, `anb.NodeConfig` should already contain the defaults. However, you're right that the code reads a little better, so I can change it back

vishalchangrani commented on follower/consensus_follower.go:
```diff
@@ -0,0 +1,108 @@
+	upstreamAccessNodeID flow.Identifier,
+	bindAddr string,
+	opts ...Option,
+) rxgo.Producer {
```
is `rxgo.Producer` a popular library? I would have preferred to just use the basic GoLang libarary to define our API. What extra does the `rxgo.Producer` gives us over simply returning a `read-only channel` ?

vishalchangrani replied:
the `rxgo` also tends to hide what is being produced (in this case `blocks`).

smnzhu replied:
I think RxGo is a pretty popular library for the observable pattern, in fact it's the official ReactiveX API for the go language: https://github.com/ReactiveX/RxGo

We definitely don't have to use it here, but I think it does a lot of the work for us, and it is very flexible from the Observer's perspective.

I only have to define the Producer function, and anyone who wants to create an Observable from it can do so while specifying various different options: https://github.com/ReactiveX/RxGo/blob/master/doc/options.md

They can create an observable that is lazy, uses a buffered channel, has a given context, [automatically does fanout](https://github.com/ReactiveX/RxGo/blob/master/README.md#connectable-observable) to multiple observers, etc.

zhangchiqing replied:
Can we bring this choice to a team discussion?

I was quite into functional programming and functional reactive programming a few years ago. And did try RxJS and other similar libraries, the experience wasn't good to be honest. For instance, I messed up the concept of hot/cold observables and made bug. The error handling was confusing especially when multiple observables are merged. Wasn't easy to debug when the consumer didn't receive an expected event. 

It does save some work, but also introduces some patterns and behaviors that not most golang developer would be familiar with. 

For instance, the "lazy and buffered" isn't clear to me what would happen if the buffered channel gets full, would the new message drop or blocking the producer? And what's beneficial being "lazy"? Does it come with a performance cost?

As this PR only introduces the Observable interfaces, it's not clear what's the full picture of how it'll be used by unstaked follower engines. What is "fanout to the multiple observers" use case? Is each observer a unstaked peer? If so, then would new staked peer connect connect and disconnect affect each other's consumption of the message? 

The original RX idea is from RxJava, and Java has a good support for generic to reuse those functions and maximize the benefit. However, in golang, with Observables, in order to provide reusable functions (like Map, ForEach), each item/message would have to be defined as `interface{}`. That means you lose type check from the compiler. For instance, we create an Observable of a stream of finalized block IDs, but the observer might cast it into block headers, then this buggy program would still compile. And it might not be easy to find out by tests, because the unittest for mismatching observable and observers could all pass. We would have to rely on integration tests to discover such mistakes.

I'm open to any choice, but just to bring up some "invisible cost" here.

uhznms replied:
Thanks Leo! I think after some discussion with @huitseeker and @vishalchangrani, we agree this probably isn't the right time to introduce this library.

However, I would love to bring this to a team discussion at some point later on to explore the idea of having a standard way (e.g perhaps RxGo library, but perhaps something else) of implementing the observable pattern.

We currently have many examples throughout the codebase where the observable pattern is being used, but differ in various ways (e.g do we use channels or callbacks? are callbacks blocking? do they arrive in order?) and I think it could perhaps be valuable to have a standard way of dealing with this pattern.

smnzhu replied:
Thanks @zhangchiqing ! I think after some discussion with @huitseeker and @vishalchangrani, we agree this probably isn't the right time to introduce this library.

However, I would love to bring this to a team discussion at some point later on to explore the idea of having a standard way (e.g perhaps RxGo library, but perhaps something else) of implementing the observable pattern.

We currently have many examples throughout the codebase where the observable pattern is being used, but differ in various ways (e.g do we use channels or callbacks? are callbacks blocking? do they arrive in order?) and I think it could perhaps be valuable to have a standard way of dealing with this pattern.

vishalchangrani commented on cmd/access/access_node_builder.go:
```diff
@@ -132,24 +147,225 @@ type FlowAccessNodeBuilder struct {
+	return builder
+}
+
+func (builder *FlowAccessNodeBuilder) BuildConsensusFollower() AccessNodeBuilder {
```
wouldn't the consensus follower of the staked AN and unstaked AN be different? The unstaked AN wouldn't have the ability to respond to sync request while the staked AN would.

smnzhu replied:
The unstaked will have the ability to respond to sync requests, eventually. The way we can disable this for now is at the networking level, ie not letting the unstaked nodes talk to each other at all. We don't need to make any changes to the engines themselves.

vishalchangrani commented on follower/consensus_follower.go:
```diff
@@ -0,0 +1,108 @@
+	anb := access.FlowAccessNode(accessNodeOptions...)
+	nodeBuilder := access.NewUnstakedAccessNodeBuilder(anb)
+
+	nodeBuilder.BuildConsensusFollower()
```
this will missout on calling `Initalize` which initializes the unstaked network.

smnzhu replied:
You're right, I forgot to add `Initialize` here :)

vishalchangrani commented on follower/consensus_follower.go:
```diff
@@ -0,0 +1,150 @@
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// ConsensusFollower is a standalone module run by third parties which provides
```
❤️  thanks for the comment and the interface

vishalchangrani commented on follower/consensus_follower.go:
```diff
@@ -0,0 +1,150 @@
+)
+
+// ConsensusFollower is a standalone module run by third parties which provides
+// a mechanism for observing the block chain. It maintains a set of subscribers
```
so correct me if I am wrong here. With this setup we can do the following:

                       | Standalone (using `main.go`)   | Library   |
 -------------+-------------------------------+---------|
  Staked AN    |               ✅                                    |    ❌       |
--------------+------------------------------- |-------- |
  Unstaked AN|              ✅                                     |    ✅       |
--------------+--------------------------------+--------+

smnzhu replied:
That's correct :)

vishalchangrani commented on follower/consensus_follower.go:
```diff
@@ -0,0 +1,150 @@
+type Config struct {
+	nodeID               flow.Identifier // the node ID of this node
+	upstreamAccessNodeID flow.Identifier // the node ID of the upstream access node
+	bindAddr             string          // address to bind on
```
fyi - will also need the upstream staked AN networking public key. I will add that here later.

smnzhu replied:
Actually, from our offline discussion my understanding was that we would only need the staked AN address? Given that, the networking public keys are exchanged when the nodes connect (which happens in the libp2p library, but I'd assume there is some way to access it),  and then the node ID can be found via the DHT?

vishalchangrani replied:
so my bad - we do need the public key because when a node in libp2p connects to the other node, it uses a [multiaddress](https://docs.libp2p.io/concepts/addressing/). The multiaddress contains the libp2p node id e.g. `/ip4/7.7.7.7/tcp/4242/p2p/**QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N**`.

The very first thing a node does after connecting to the other node is a TLS [handshake to create an encrypted connection](https://github.com/libp2p/specs/blob/master/tls/tls.md#libp2p-public-key-extension).
For this to work, the node initiating the connection must have the public key of the peer. From this public key it can derive the libp2p node id and verify that the node id of the peer matches.

smnzhu replied:
@vishalchangrani okay interesting, so to check my understanding basically the libp2p node id is just a hash of the public key right?

smnzhu replied:
and is the Flow node ID different from the libp2p node id?

vishalchangrani replied:
> @vishalchangrani okay interesting, so to check my understanding basically the libp2p node id is just a hash of the public key right?

correct..it is just a hash of the public key.

vishalchangrani replied:
> and is the Flow node ID different from the libp2p node id? If it is the same, then we would _only_ need to provide the node id since upon connection to a remote peer, the peer will provide the public key for you to verify right? https://github.com/libp2p/specs/blob/master/tls/tls.md#:~:text=Extension-,In%20order%20to%20prove,the%20private%20host%20key,-The

The Flow node id is different from the libp2p node id. The Flow node id is also a hash of the public key but its a different hash AND it may not be a hash in the future and could be something else.

smnzhu replied:
Okay, so to summarize, we do need all three of the following:
* Staked node's Flow 
* Staked node's public key (to verify upon connection that it is who it says it is)
* Staked node IP address (to connect to it in the first place)

Is this correct?

huitseeker replied:
@smnzhu So this should help:
https://github.com/onflow/flow-go/blob/6c4685b73e99206b199bd3c5ccc7dd7ced7c8517/network/p2p/keyTranslator.go#L34-L100

https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#peer-ids

i.e.  Flow's Network Public Keys and LibP2P's public key are a marshalling standard away from each other and inter-convertible. 

The libp2p PeerID is a hash of the libp2p public key. BUT when that public key is less than 42 bytes (which is the case for the ECC keys we use, at 32 bytes), that hash is the identity, after which the key is base58 encoded.

Hence libp2p can [tell you the key for the node you have the ID](https://github.com/libp2p/go-libp2p-core/blob/master/peer/peer.go#L91).

So if "˜=" means equal modulo format change:

Flow network key ˜= libp2p Public key ˜= libp2p PeerID

Besides that, Flow's (internal) NodeID is a SHA3-256 hash of the (Flow) staking public key, as you can see here:
https://github.com/onflow/flow-go/blob/64a592f3a099d2a313c23cdeedc3eba2d7f09538/cmd/bootstrap/cmd/keys.go#L69

This should probably be in a doc somewhere (@vishalchangrani  if you have suggestions as to where to put this?)

vishalchangrani replied:
It will be best to include this as a code comment [here](https://github.com/onflow/flow-go/blob/master/network/p2p/keyTranslator.go).

vishalchangrani reviewed (approved):
lgtm!

smnzhu commented:
@vishalchangrani some more changes since your last review, I had to do quite a bit of refactoring for the following reason:

The `consensus_follower` package needs to import the `AccessNodeBuilder`. Previously, this was defined in the [`main` package](https://github.com/onflow/flow-go/pull/1085/files#diff-8d9f5e7f5c389eaf47e11124c15e78a38eb3f941db0cc46529d95decbc7d4ef2R1),  where `consensus_follower` was [importing](https://github.com/onflow/flow-go/pull/1085/files/106a046e3dda73ae8474f9952a65b4cf5c7269ab#diff-d85a41646d61a407b2e2fcd0c891da861eecb57e4314ff17f4bdc98bb8de0692R9) it from.

However, this is [not allowed](https://stackoverflow.com/a/44421114) in go, so I had to move those into a separate package. This caused some compilation issues because the fields in [`AccessNodeConfig`](https://github.com/onflow/flow-go/pull/1085/files#diff-ced29c0a3c7e4ab8ea075969455e282214292b980d7f16eee57a94beb5976bc8R84-R103) are mostly private, but they were being accessed in `main.go`. Once `access_node_builder` was moved to a separate package, `main.go` no longer had access to these fields.

I decided to move most of the code that builds the access node out of `main.go` and into the new `node_builder` package as well. Most of this goes inside the newly defined `Build()` method on the `AccessNodeBuilder` interface.

A couple other minor changes:
* I added a `metricsEnabled` field since we want to be able to toggle it off for the unstaked library. Please check that the implementation of this field is comprehensive and I am catching all the places where components relevant to metrics are initialized
* I changed some code to make sure that the `unicaseMessageTimeout` config field is actually propagated down to the `initMiddleware` method.

Please review the new changes since your last review, thanks!

vishalchangrani commented:
> @vishalchangrani some more changes since your last review, I had to do quite a bit of refactoring for the following reason:
> 
> The `consensus_follower` package needs to import the `AccessNodeBuilder`. Previously, this was defined in the [`main` package](https://github.com/onflow/flow-go/pull/1085/files#diff-8d9f5e7f5c389eaf47e11124c15e78a38eb3f941db0cc46529d95decbc7d4ef2R1), where `consensus_follower` was [importing](https://github.com/onflow/flow-go/pull/1085/files/106a046e3dda73ae8474f9952a65b4cf5c7269ab#diff-d85a41646d61a407b2e2fcd0c891da861eecb57e4314ff17f4bdc98bb8de0692R9) it from.
> 
> However, this is [not allowed](https://stackoverflow.com/a/44421114) in go, so I had to move those into a separate package. This caused some compilation issues because the fields in [`AccessNodeConfig`](https://github.com/onflow/flow-go/pull/1085/files#diff-ced29c0a3c7e4ab8ea075969455e282214292b980d7f16eee57a94beb5976bc8R84-R103) are mostly private, but they were being accessed in `main.go`. Once `access_node_builder` was moved to a separate package, `main.go` no longer had access to these fields.
> 
> I decided to move most of the code that builds the access node out of `main.go` and into the new `node_builder` package as well. Most of this goes inside the newly defined `Build()` method on the `AccessNodeBuilder` interface.
> 
> A couple other minor changes:
> 
> * I added a `metricsEnabled` field since we want to be able to toggle it off for the unstaked library. Please check that the implementation of this field is comprehensive and I am catching all the places where components relevant to metrics are initialized
> * I changed some code to make sure that the `unicastMessageTimeout` config field is actually propagated down to the `initMiddleware` method.
> 
> Please review the new changes since your last review, thanks!

Then you may have already completed this [issue](https://github.com/dapperlabs/flow-go/issues/5688 - I was asked by Jordan to consolidate all the node builder stuff into its own separate package)

smnzhu commented:
@vishalchangrani Thanks, I'll add that to the PR summary :)

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -293,22 +294,32 @@ func (fnb *FlowNodeBuilder) initMetrics() {
-		Compliance: metrics.NewComplianceCollector(),
-		Cache:      metrics.NewCacheCollector(fnb.RootChainID),
-		Mempool:    mempools,
+		Network:    metrics.NewNoopCollector(),
```
✅

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -293,22 +294,32 @@ func (fnb *FlowNodeBuilder) initMetrics() {
+		Cache:      metrics.NewNoopCollector(),
+		Mempool:    metrics.NewNoopCollector(),
 	}
+	if fnb.BaseConfig.metricsEnabled {
```
yup, this makes sense to me.

huitseeker reviewed (approved):
Nice work @smnzhu !

huitseeker commented:
/cc @awfm9 @Ullaakut
